divi/src/kernel.cpp

```diff
@@ -560,6 +564,48 @@ bool CheckProofOfStake(const CBlock& block, int blockHeight, uint256& hashProofO
     return true;
 }
 
+bool CheckCoinstakeForVaults(const CTransaction& tx, const CBlockRewards& expectedRewards,
+                             const CCoinsViewCache& view)
+{
+    if (!tx.IsCoinStake())
+        return error("%s: transaction is not a coinstake", __func__);
+
+    CAmount nValueIn = 0;
+    bool foundVault = false;
+    CScript vaultScript;
+    for (const auto& in : tx.vin) {
+        const auto& prevOut = view.GetOutputFor(in);
+        nValueIn += prevOut.nValue;
+        if (!IsStakingVaultScript(prevOut.scriptPubKey))
+            continue;
+
+        if (foundVault) {
+            /* CheckProofOfStake already verifies that all inputs used are
+               from a single script.  */
+            assert(vaultScript == prevOut.scriptPubKey);
```

Comment:
We could.  However, easing the requirements later would be a hard fork by itself, so it cannot be easily done anyway.  And with that in mind, I personally think it is actually better to more strictly enforce the assumptions of the code, rather than silently "just do something".

But if you prefer to replace the `assert` here with a `return error`, I can of course do that.